### ISSUE TITLE
FD leak in queue tests

### ISSUE THREAD

**Issue by divergentdave** [divviup/janus#3405]
I'm running into test failures when running the test suite locally. I see "Too many open files" first, then a series of "metrics provider already shut down" messages. I bisected this to #3369, then bisected to 1bc432aaf540e313081f569cab8867e13548c0d7 within that. Interestingly, this is passing just fine on CI machines, so the number of threads and/or default ulimit settings must play a factor. Running one quickcheck test is not enough to trigger the issue, but running all four is.

I tried stracing these tests, and it looks like it fails after using 1024 file descriptors. I think about half are attributable to `fcntl(fd, F_DUPFD_CLOEXEC, 3)` calls, and about half are attributable to `eventfd2(0, EFD_CLOEXEC|EFD_NONBLOCK)`, though it's hard to follow FDs from opening to closing just based on logs. The `/proc/$PID/fd` listing shows a lot of `anon_inode:[eventfd]` and `anon_inode:[eventpoll]` plus some sockets.

**Comment by divergentdave** [comment]
I tried replacing the `mem::forget()` of a `CancelDropGuard` with a flag to disable sending a cancellation message. This appears to have fixed the issue. Making this change means we no longer leak a sender of the message channel. (since `forget()` both skips our `Drop` implementation and forgets the struct's fields) The stack traces of the `eventfd2()` and `fcntl()` calls mentioned above come from mio, and thus from Tokio. Looking at the [Tokio documentation on runtime shutdown](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html), it says that when we drop the `Runtime` (which is in a temporary), that will cancel any spawned tasks. Notably, the dispatcher will run until this cancellation, because it normally exits when all message channel senders are dropped, but those are currently leaked on the happy path. The documentation also says that dropping `Runtime` will block until all tasks have yielded, and the `Runtime` won't be usable after. However, I wonder if some component of the runtime, like a reactor or I/O driver, may be kept alive by the leaked channel sender, after the drop call is done. Perhaps this could in turn cause this side effect of hanging on to a few FDs.

**Comment by divergentdave** [comment]
I ran this under DHAT, to look at allocations that leaked through the end of the program, and thus get hints at what runtime components are sticking around. Running under Valgrind, it terminated early with an OpenTelemetry error ("send failed because receiver is gone"), likely due to a shutdown race condition. This means it didn't get to the point of FD exhaustion, but it still ran enough times for me to get some data. There was one FD-related allocation I saw, which had 91 allocations live at shutdown. The stack trace included `tokio::runtime::builder::Builder::build_threaded_runtime`, `tokio::runtime::io::driver::Driver::new`, `mio::event::events::Events::with_capacity`, and `with_capacity_in<libc::unix::linux_like::epoll_event>`. I think this is pretty good evidence that parts of the runtime are being kept alive after we drop the runtime itself, and they are in turn hanging on to open file descriptors.
